### PR TITLE
166 invalid resources

### PR DESCRIPTION
**Pull request by fheuschkel**
Still work in progress.
TODO: 
- change from /fhir-spec/valid? to  modified /fhir-spec/explain-data in all useful situations. 
- maybe add more unit tests
What do you think so far?

### CODE REVIEW

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 1]
```diff
@@ -46,16 +47,99 @@
...
+       [spec-error (when-let
+                     [spec (s2/get-spec (keyword "fhir" type))]
+                     (s2/explain-data spec resource))]
+       (assoc
+         spec-error
+         :fhir/issues
+         (into
+           []
+           (map (fn [problem]
+                  (let [{:keys [val via in pred]} problem]
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:126]: just move the destructuring into the `fn` form

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 2]
```diff
@@ -46,16 +47,99 @@
...
+                     severity
+                     :fhir.issues/code
+                     "invariant"
+                     :fhir.issues/diagnostics
+                     (get-diagnostics-from-problem pred val via)
+                     :fhir.issues/expression
+                     (fhir-path in resource)}))
+                (:clojure.spec.alpha/problems spec-error)))))
+     {:fhir/issues
+      (conj
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:137]: you can just write a literal vector with a map in it. no need for `conj` here

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 3]
```diff
@@ -46,16 +47,99 @@
...
+        []
+        {:fhir.issues/severity "error"
+         :fhir.issues/code "value"
+         :fhir.issues/diagnostics
+         "Given resource does not contain :resourceType key!"})})))
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:142]: Please output the same message, that would be output if another property would be missing. We like to never expose internal things like keywords.

File: modules/fhir-structure/test/blaze/fhir/spec_test.clj [hunk 4]
```diff
@@ -58,13 +59,45 @@
...
     (are [resource] (not (fhir-spec/valid? resource))
+      {}
       {:resourceType "Patient"
        :id ""}
       {:resourceType "Patient"
        :id "/"})))
 
+(deftest explain-data
+  (testing "valid resources"
+    (are [resource] (not (fhir-spec/explain-data resource))
```
Comment (alexanderkiel) [inline modules/fhir-structure/test/blaze/fhir/spec_test.clj:73]: not -> nil?

File: modules/fhir-structure/test/blaze/fhir/spec_test.clj [hunk 5]
```diff
@@ -58,13 +59,45 @@
...
+      [:fhir/issues 0 :fhir.issues/code] := "invariant"
+      [:fhir/issues 0 :fhir.issues/diagnostics] :=
+      "Error on Value ``. Expected type is `code`,
+      regex `[^\\s]+(\\s[^\\s]+)*`."
+      [:fhir/issues 0 :fhir.issues/expression] := "name[0].use"
+      [:fhir/issues 1 :fhir.issues/severity] := "error"
+      [:fhir/issues 1 :fhir.issues/code] := "invariant"
+      [:fhir/issues 1 :fhir.issues/diagnostics] :=
+      "Error on Value `[]`. Expected type is `string`."
+      [:fhir/issues 1 :fhir.issues/expression] := "name[0].text")))
```
Comment (alexanderkiel) [inline modules/fhir-structure/test/blaze/fhir/spec_test.clj:99]: please add tests of invalid backbone elements and non-primitive elements like HumanName or Quantity

File: modules/rest-util/src/blaze/handler/util.clj [hunk 6]
```diff
@@ -56,30 +56,52 @@
...
-      (assoc :expression [expression]))]})
+  (if issues
+    {:resourceType "OperationOutcome"
+     :issue
+     (into
+       []
+       (map (fn [problem]
+              (let [{:fhir.issues/keys [severity code diagnostics expression]}
+                    problem]
+                (cond->
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/handler/util.clj:71]: please do not duplicate this code here, use a function

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 7]
```diff
@@ -3,7 +3,8 @@
     [blaze.fhir.spec.impl]
     [clojure.alpha.spec :as s2]
     [clojure.spec.alpha :as s]
-    [clojure.string :as str]))
+    [clojure.string :as str])
+  (:import (java.util.regex Pattern)))
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:7]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 8]
```diff
@@ -53,11 +54,92 @@
     false))
 
 
+(defn- fhir-path-data
+  "Given a vector `elem` and some `data` structure containing maps and vectors,
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:58]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 9]
```diff
@@ -53,11 +54,92 @@
...
+  for witch the entries in elem correspond to the keys and vector entries of the
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:59]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 10]
```diff
@@ -53,11 +54,92 @@
...
+  given data, returns a vector containing keys and index positions of the vector
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:60]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 11]
```diff
@@ -46,16 +47,99 @@
...
+  {:arglists '([elem data])}
+  [[elem & elems] data]
+  (if (and (map? elem) (vector? data))
+    (cons (first (keep-indexed #(when (= %2 elem) %1) data))
+          (when elems (fhir-path-data elems elem)))
+    (cons elem (if elems (fhir-path-data elems (get data elem)) []))))
+
+
+(defn fhir-path
+  "Given a vector `path` and a `resource`, returns the fhir-path as String for
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:71]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj (hunk 2 above)
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:137]: Done

File: modules/fhir-structure/test/blaze/fhir/spec_test.clj (hunk 4 above)
Comment (fheuschkel) [reply modules/fhir-structure/test/blaze/fhir/spec_test.clj:73]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 12]
```diff
@@ -46,16 +47,99 @@
...
+  the given path."
+  [path resource]
+  (subs (reduce (fn [x y] (str x (if (keyword? y)
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:74]:
```
(str/join "." (map #(if (keyword? %)
                        (name %)
                        (format "[%s]" %))
                     (fhir-path-data path resource))) 
```
Ok like this? Did not use partition-by...

[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:74]:
```
(->>(fhir-path-data path resource)
    (reduce (fn [list elem] (if (keyword? elem)
                            (conj list [(name elem)])
                            (assoc
                              list
                              (- (count list) 1)
                              (conj
                                (last list)
                                (format "[%d]" elem)))))
          [])
      (map #(str/join "" %))
      (str/join "."))
```
What do you think about that?

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 13]
```diff
@@ -46,16 +47,99 @@
...
+                                   (format ".%s" (name y))
+                                   (format "[%s]" y)))) ""
+                (fhir-path-data path resource)) 1))
+
+
+(defn get-regex
+  "Returns the first regex pattern in `form`."
+  [form]
+  (clojure.walk/postwalk
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:83]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 14]
```diff
@@ -46,16 +47,99 @@
...
+    #(if (sequential? %)
+       (some identity %)
+       (when (= (type %) Pattern) %))
+    form))
+
+
+(defn get-diagnostics-from-problem
+  "Returns a diagnostics message from the given arguments pred, val and via.
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:91]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 15]
```diff
@@ -46,16 +47,99 @@
...
+  pred: A Function to evaluate the given val.
+  val: Error Value.
+  via: Vector-path of fhir types to the given error.
+  If pred contains Regex-patetrns to check, the first is included in diagnostics
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:95]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 16]
```diff
@@ -46,16 +47,99 @@
...
+  message."
+  [pred val via]
+  (if-let [regex (get-regex pred)]
+    (format "Error on Value `%s`. Expected type is `%s`, regex `%s`."
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:99]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 17]
```diff
@@ -46,16 +47,99 @@
...
-    (s2/explain-data spec resource)))
+  "Given a resource witch includes :resourceType key and severity as string,
+  returns nil if the resource is conform to the spec :fhir/resourceType,
+  else a map with at least the keys :fhir/issues whoes value is a collection of
+  issue-maps to build a OperationsOutcome and ::problems whose value is a
+  collection of problem-maps, where problem-map has at least :path :pred and
+  :val keys describing the predicate and the value that failed at that path."
+  {:arglists '([resource] [resource severity])}
+  ([{type :resourceType :as resource}] (explain-data resource "error"))
+  ([{type :resourceType :as resource} severity]
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:114]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 18]
```diff
@@ -46,16 +47,99 @@
...
+    (format "Error on Value `%s`. Expected type is `%s`."
+            val (name (last via)))))
+
+
 (defn explain-data
-  {:arglists '([resource])}
-  [{type :resourceType :as resource}]
-  (when-let [spec (s2/get-spec (keyword "fhir" type))]
-    (s2/explain-data spec resource)))
+  "Given a resource witch includes :resourceType key and severity as string,
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:106]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj (hunk 1 above)
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:126]: Done

File: modules/fhir-structure/test/blaze/fhir/spec_test.clj (hunk 5 above)
Comment (fheuschkel) [reply modules/fhir-structure/test/blaze/fhir/spec_test.clj:99]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 19]
```diff
@@ -46,16 +47,99 @@
...
+  "Returns a diagnostics message from the given arguments pred, val and via.
+  pred: A Function to evaluate the given val.
+  val: Error Value.
+  via: Vector-path of fhir types to the given error.
+  If pred contains Regex-patetrns to check, the first is included in diagnostics
+  message."
+  [pred val via]
+  (if-let [regex (get-regex pred)]
+    (format "Error on Value `%s`. Expected type is `%s`, regex `%s`."
+            val (name (last via)) regex)
```
[the remark this answers is not in the document]
Comment (fheuschkel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:100]: Done

File: modules/fhir-structure/src/blaze/fhir/spec.clj (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexanderkiel) [reply modules/fhir-structure/src/blaze/fhir/spec.clj:74]: `(update list (dec (count list)) str (format "[%d]" elem))`

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 20]
```diff
@@ -54,64 +56,82 @@
     false))
 
 
-(defn- fhir-path-data
-  "Given a vector `elem` and some `data` structure containing maps and vectors,
-  for witch the entries in elem correspond to the keys and vector entries of the
-  given data, returns a vector containing keys and index positions of the vector
-  elements for the given data."
-  {:arglists '([elem data])}
+(defn fhir-path-data
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:59]: private

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 21]
```diff
@@ -54,64 +56,82 @@
...
 (defn fhir-path
-  "Given a vector `path` and a `resource`, returns the fhir-path as String for
+  "Given a vector `path` and a `resource`, returns the fhir-path as string for
   the given path."
   [path resource]
-  (subs (reduce (fn [x y] (str x (if (keyword? y)
-                                   (format ".%s" (name y))
-                                   (format "[%s]" y)))) ""
-                (fhir-path-data path resource)) 1))
+  (->>(fhir-path-data path resource)
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:76]: formatting (space between ->> and ()

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 22]
```diff
@@ -54,64 +56,82 @@
...
-(defn get-diagnostics-from-problem
-  "Returns a diagnostics message from the given arguments pred, val and via.
-  pred: A Function to evaluate the given val.
-  val: Error Value.
-  via: Vector-path of fhir types to the given error.
-  If pred contains Regex-patetrns to check, the first is included in diagnostics
-  message."
+(defn- extended-fhir-namespace?
+  "Returns the namesake part and name of `key` if the namespace of `key` is more
+  than :fhir. Otherwise returns nil."
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:99]: a namespace is a string so no : here

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 23]
```diff
@@ -54,64 +56,82 @@
...
+  [key]
+  (when (str/starts-with?  (str key) ":fhir")
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:101]: `(str/starts-with? (namespace key) "fhir")`

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 24]
```diff
@@ -54,64 +56,82 @@
...
+    (str/replace (subs (str key) 6) "/" ".")))
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:102]: please user split and join here

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 25]
```diff
@@ -54,64 +56,82 @@
...
+
+
+(defn- get-diagnostics-from-problem
+  "Returns a diagnostics message from the given arguments `pred`, `val` and `via`.
+  `pred`: A Function to evaluate the given `val`.
+  `val`: Error Value.
+  `via`: Vector-path of fhir types to the given error.
+  If `pred` contains regex-patterns to check, the first one is included in
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:110]: newline before

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 26]
```diff
@@ -54,64 +56,82 @@
...
+  diagnostics message."
   [pred val via]
   (if-let [regex (get-regex pred)]
-    (format "Error on Value `%s`. Expected type is `%s`, regex `%s`."
+    (format "Error on value `%s`. Expected type is `%s`, regex `%s`."
             val (name (last via)) regex)
-    (format "Error on Value `%s`. Expected type is `%s`."
-            val (name (last via)))))
+    (if (= (str pred)  "clojure.core/coll?")
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:116]: (= `coll? pred)

Pred is a symbol and you can write a literal symbol in source code by quoting it. the back tics syntax quote will automatically resolve the symbol according the namespace declarations.

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 27]
```diff
@@ -54,64 +56,82 @@
...
 
 
-(defn get-diagnostics-from-problem
-  "Returns a diagnostics message from the given arguments pred, val and via.
-  pred: A Function to evaluate the given val.
-  val: Error Value.
-  via: Vector-path of fhir types to the given error.
-  If pred contains Regex-patetrns to check, the first is included in diagnostics
-  message."
+(defn- extended-fhir-namespace?
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:97]: the function doesn't return boolean so no ? here

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 28]
```diff
@@ -54,64 +56,82 @@
...
   (if-let [regex (get-regex pred)]
-    (format "Error on Value `%s`. Expected type is `%s`, regex `%s`."
+    (format "Error on value `%s`. Expected type is `%s`, regex `%s`."
             val (name (last via)) regex)
-    (format "Error on Value `%s`. Expected type is `%s`."
-            val (name (last via)))))
+    (if (= (str pred)  "clojure.core/coll?")
+      (format "Error on value `%s`. Expected type is `json-array`."
+              val)
+      (->> (if-let [namespace (extended-fhir-namespace? (last via))]
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:119]: I would just return the right thing from a renamed `extended-shir-namespace?` here

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 29]
```diff
@@ -54,64 +56,82 @@
...
+             namespace
+             (name (last via)))
+           (format "Error on value `%s`. Expected type is `%s`."
+               val )))))
 
 
 (defn explain-data
-  "Given a resource witch includes :resourceType key and severity as string,
+  "Given a resource, witch includes :resourceType key and severity as string,
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:127]: no longer takes severity
which

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 30]
```diff
@@ -122,24 +142,21 @@
...
+           (map (fn [{:keys [val via in pred]}]
+                  {:fhir.issues/severity
+                   "error"
+                   :fhir.issues/code
+                   "invariant"
+                   :fhir.issues/diagnostics
+                   (get-diagnostics-from-problem pred val via)
+                   :fhir.issues/expression
+                   (fhir-path in resource)})
                 (:clojure.spec.alpha/problems spec-error)))))
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:154]: ::s/problems

File: modules/fhir-structure/src/blaze/fhir/spec.clj [hunk 31]
```diff
@@ -122,24 +142,21 @@
          :fhir/issues
          (into
```
Comment (alexanderkiel) [inline modules/fhir-structure/src/blaze/fhir/spec.clj:143]: please put the generation of fhir issues into it's own function

File: modules/fhir-structure/test/blaze/fhir/spec_test.clj [hunk 32]
```diff
@@ -68,9 +67,22 @@
       {:resourceType "Patient"
        :id "/"})))
 
+
+(deftest fhir-path
```
Comment (alexanderkiel) [inline modules/fhir-structure/test/blaze/fhir/spec_test.clj:71]: please add a test were the index isn't zero

File: modules/fhir-structure/test/blaze/fhir/spec_test.clj [hunk 33]
```diff
@@ -80,23 +92,57 @@
...
+      "Error on value `[]`. Expected type is `string`."
+      [:fhir/issues 1 :fhir.issues/expression] := "name[0].text"))
+
+  (testing "invalid backbone-element"
+    (given (fhir-spec/explain-data {:resourceType "Patient"
+                                    :contact ""})
+      [:fhir/issues 0 :fhir.issues/severity] := "error"
+      [:fhir/issues 0 :fhir.issues/code] := "invariant"
+      [:fhir/issues 0 :fhir.issues/diagnostics] :=
+      "Error on value ``. Expected type is `json-array`."
```
Comment (alexanderkiel) [inline modules/fhir-structure/test/blaze/fhir/spec_test.clj:117]: `` `json-array` `` -> JSON array

File: modules/fhir-structure/test/blaze/fhir/spec_test.clj [hunk 34]
```diff
@@ -80,23 +92,57 @@
...
+  (testing "invalid non-primitive element "
+    (given (fhir-spec/explain-data {:resourceType "Patient"
+                                    :name ""})
+      [:fhir/issues 0 :fhir.issues/severity] := "error"
+      [:fhir/issues 0 :fhir.issues/code] := "invariant"
+      [:fhir/issues 0 :fhir.issues/diagnostics] :=
+      "Error on value ``. Expected type is `json-array`."
+      [:fhir/issues 0 :fhir.issues/expression] := "name"))
+
+  (testing "Include namesake part if more than fhir?"
```
Comment (alexanderkiel) [inline modules/fhir-structure/test/blaze/fhir/spec_test.clj:129]: namesake -> namespace
remove the question mark

File: modules/rest-util/src/blaze/handler/util.clj [hunk 35]
```diff
@@ -55,53 +55,59 @@
...
+           severity
+           (assoc :severity severity)
+           diagnostics
+           (assoc :diagnostics diagnostics)
+           (coll? expression)
+           (assoc :expression expression)
+           (and (not (coll? expression))
+                (some? expression))
+           (assoc :expression [expression])
+           )) issues))
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/handler/util.clj:72]: formatting

File: modules/rest-util/src/blaze/handler/util.clj [hunk 36]
```diff
@@ -55,53 +55,59 @@
     "exception"))
 
 
-(defn operation-outcome
-  [{:fhir/keys [issues issue operation-outcome]
+(defn- operation-outcome-issues [issues category]
+  (map (fn [{:fhir.issues/keys [severity code diagnostics expression]}]
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/handler/util.clj:59]: mapv

File: modules/rest-util/src/blaze/handler/util.clj [hunk 37]
```diff
@@ -55,53 +55,59 @@
...
+     (assoc :expression [expression]))])
+
+
+(defn operation-outcome
+  [{:fhir/keys [issues]
+    ::anom/keys [category] :as data}]
   (if issues
     {:resourceType "OperationOutcome"
      :issue
      (into
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/handler/util.clj:106]: you can remove the Ito here

File: modules/rest-util/src/blaze/handler/util.clj [hunk 38]
```diff
@@ -55,53 +55,59 @@
...
+           (assoc :expression [expression])
+           )) issues))
+
+
+(defn- operation-outcome-issue
+  [{:fhir/keys [issue operation-outcome]
     :fhir.issue/keys [expression]
     :blaze/keys [stacktrace]
     ::anom/keys [category message]}]
+  [(cond->
```
Comment (alexanderkiel) [inline modules/rest-util/src/blaze/handler/util.clj:80]: I would move the vector out, because the method is in singular ans do should only return one issue
